core/src/main/java/io/parsingdata/metal/data/selection/ByName.java

```diff
@@ -56,20 +60,21 @@ public static ParseValue getValue(final ParseGraph graph, final String name) {
     public static ImmutableList<Value> getAllValues(final ParseGraph graph, final String name) {
         checkNotNull(graph, "graph");
         checkNotNull(name, "name");
-        return getAllValuesRecursive(graph, name);
+        return reverse(getAllValuesRecursive(ImmutableList.create(graph), new ImmutableList<>(), name).computeResult());
     }
 
-    private static ImmutableList<Value> getAllValuesRecursive(final ParseGraph graph, final String name) {
-        if (graph.isEmpty()) { return new ImmutableList<>(); }
-        final ImmutableList<Value> tailResults = getAllValuesRecursive(graph.tail, name);
+    private static SafeTrampoline<ImmutableList<Value>> getAllValuesRecursive(final ImmutableList<ParseGraph> graphList, final ImmutableList<Value> values, final String name) {
+        if (graphList.isEmpty()) { return complete(() -> values); }
+        final ParseGraph graph = graphList.head;
+        if (graph.isEmpty()) { return intermediate(() -> getAllValuesRecursive(graphList.tail, values, name)); }
         final ParseItem head = graph.head;
         if (head.isValue() && head.asValue().matches(name)) {
-            return tailResults.add(head.asValue());
+            return intermediate(() -> getAllValuesRecursive(graphList.tail.add(graph.tail), values.add(head.asValue()), name));
         }
         if (head.isGraph()) {
-            return tailResults.add(getAllValuesRecursive(head.asGraph(), name));
+            return intermediate(() -> getAllValuesRecursive(graphList.tail.add(graph.tail).add(graph.head.asGraph()), values, name));
         }
-        return tailResults;
+        return intermediate(() -> getAllValuesRecursive(graphList.tail.add(graph.tail), values, name));
```

Comment:
Wow indeed! Implemented! I named the methods `addIfGraph()` and `addIfMatchingValue()` to express what they do.